### DISCUSSION TITLE
UnsupportedChildType when using formsets?

### DISCUSSION THREAD

**Discussion by kimutaiRop** [jazzband/django-polymorphic#812] [Q&A]
I have a polymorphic model that I made and it has many child models which I want the users to be choosing dynamically, posting the first product has no issue, I was able to select the form to send to the user, however when I try subsequent queries for the model to send to user I get error

`polymorphic.formsets.models.UnsupportedChildType: The 'ShirtsAndBlouseForm' found a 'PhoneTablets' model in the queryset, but no form class is registered to display it.`
my forms look like this

```
PhonesTabsForm = polymorphic_modelformset_factory(Product, fields=product_fields, formset_children=(
    PolymorphicFormSetChild(PhoneTablets, fields=product_fields + ['screen_size', 'battery_type',
                                                                   'battery_capacity', 'ram_measure', 'ram_storage',
                                                                   'internal_storage']),
))
```

```
ShirtsAndBlouseForm = polymorphic_modelformset_factory(Product, fields=product_fields, formset_children=(
    PolymorphicFormSetChild(ShirtsAndBlouse, fields=product_fields + ['techniques',
                                                                      'fit_type', 'length', 'material', 'fabric_type',
                                                                      'wear_type', 'design'], ),
))
```
I dont understand what could be happening that it can allow me to select the form during the first time but not the second time when product exist in the database

**Comment by lukatavcer** [comment]
I had the same issue as I forgot to include the child class (StackedPolymorphicInline.Child) in the inline (StackedPolymorphicInline).

**Comment by JohananOppongAmoateng** [comment]
@bckohan i think this is a user error not a bug in the library. I believe we can close this

**Comment by bckohan** [comment]
@kimutaiRop 

That error is basically saying:

> you instantiated **`ShirtsAndBlouseForm`**, but the **queryset you gave it (or the default one it uses)** contains at least one **`PhoneTablets`** object — and `ShirtsAndBlouseForm` only knows how to render `ShirtsAndBlouse`, so it blows up.

Why it “worked the first time”
On your first request there were **no `PhoneTablets` rows yet** (or no rows of other child types), so the default queryset happened to contain only the type your formset knows about. Once you saved a `PhoneTablets`, the next time you create `ShirtsAndBlouseForm` with the **default queryset**, it now includes that `PhoneTablets` row → `UnsupportedChildType`. This is expected behavior for [django-polymorphic formsets.](https://django-polymorphic.readthedocs.io/en/latest/formsets.html)

## The fix: always filter the queryset to the child type for that formset

Even though you create the factory like this, **the filtering is not “baked in”**:

```python
ShirtsAndBlouseForm = polymorphic_modelformset_factory(
    Product,
    fields=product_fields,
    formset_children=(
        PolymorphicFormSetChild(ShirtsAndBlouse, fields=product_fields + [...]),
    )
)
```

You must do this when *instantiating* the formset:

```python
# GET
formset = ShirtsAndBlouseForm(
    queryset=Product.objects.instance_of(ShirtsAndBlouse)
)

# POST (must use the same queryset concept!)
formset = ShirtsAndBlouseForm(
    request.POST,
    request.FILES,
    queryset=Product.objects.instance_of(ShirtsAndBlouse)
)
```

And similarly:

```python
formset = PhonesTabsForm(
    queryset=Product.objects.instance_of(PhoneTablets)
)
```

`instance_of()` is the standard django-polymorphic queryset filter for child types.

### If you’re using the formset only to create *new* items

Use an empty queryset so it never tries to render existing objects of other types:

```python
formset = ShirtsAndBlouseForm(queryset=Product.objects.none())
```

## Alternative: use *one* formset that registers *all* child types

If your UI is “one page that can edit mixed products”, then you should register *all* children in a single formset, and only *display* the subset you want in the template (or via initial client-side filtering). The library’s rule is: **every model instance in the queryset must have a registered child form**.

---

### Here's a checklist

* Separate formset per product category? **Filter queryset per category** (`instance_of(...)`).
* Create-only flow? Use `queryset=Product.objects.none()`.
* Mixed edit list? Use a **single** polymorphic formset registering all child types.
